File: codigo/soluciones/kat/beehouseperimeter.cpp
#include <bits/stdc++.h>
using namespace std;
#define mp(a,b) make_pair(a, b)
int ffill(vector<vector<int>> &grid, const int &i, const int &j, const int &N) {
    grid[i][j] = -1;
    int ans = 0;
    for (int d = -1; d <= 1; d += 2) {
        if (i + d >= 0 && i + d < N) {
            if (grid[i + d][j] == 0) {
                ans += ffill(grid, i + d, j, N);
            }else if (grid[i + d][j] == 1) {
                ++ans;
            }
        }
        if (j + d >= 0 && j + d < N) {
            if (grid[i][j + d] == 0) {
                ans += ffill(grid, i, j + d, N);
            }else if (grid[i][j + d] == 1) {
                ++ans;
            }
        }
        if (i + d >= 0 && i + d < N && j + d >= 0 && j + d < N) {
            if (grid[i + d][j + d] == 0) {
                ans += ffill(grid, i + d, j + d, N);
            }else if (grid[i + d][j + d] == 1) {
                ++ans;
            }
        }
    }
    return ans;
}
signed main() {
    cin.sync_with_stdio(false);
    cin.tie(nullptr);
    int R, K;
    cin >> R >> K;
    vector<vector<int>> grid(2 * R + 1, vector<int>(2 * R + 1));
    unordered_map<int, pair<int, int>> pos;
    int aux = 1;
    for (int i = 0; i < R; ++i) {
        for (int j = 0; j < R + i; ++j) {
            pos[aux++] = mp(i + 1, j + 1);
        }
    }
    for (int i = R; i < 2 * R - 1; ++i) {
        for (int j = i - R + 1; j < 2 * R - 1; ++j) {
            pos[aux++] = mp(i + 1, j + 1);
        }
    }
    while (K--) {
        int idx;
        cin >> idx;
        grid[pos[idx].first][pos[idx].second] = 1;
    }
    cout << ffill(grid, 0, 0, 2 * R + 1) << '\n';
    cout.flush();
    return 0;
}

File: codigo/soluciones/kat/beehouseperimeter.py
import sys
sys.setrecursionlimit(1000000)
def dfs(grid, i, j, N):
    grid[i][j] = -1
    ans = 0
    for d in [-1, 1]:
        if 0 <= i + d < N:
            if grid[i + d][j] == 0:
                ans += dfs(grid, i + d, j, N)
            elif grid[i + d][j] == 1:
                ans += 1
        if 0 <= j + d < N:
            if grid[i][j + d] == 0:
                ans += dfs(grid, i, j + d, N)
            elif grid[i][j + d] == 1:
                ans += 1
        if 0 <= i + d < N and 0 <= j + d < N:
            if grid[i + d][j + d] == 0:
                ans += dfs(grid, i + d, j + d, N)
            elif grid[i + d][j + d] == 1:
                ans += 1
    return ans
R, K = list(map(int, input().split()))
grid = [[0 for _ in range(2 * R + 1)] for _ in range(2 * R + 1)]
pos = [0 for _ in range(R**3 - (R - 1)**3)]
aux = 0
for i in range(R):
    for j in range(R + i):
        pos[aux] = (i + 1, j + 1)
        aux += 1
for i in range(R, 2 * R - 1):
    for j in range(i - R + 1, 2 * R - 1):
        pos[aux] = (i + 1, j + 1)
        aux += 1
idxs = list(map(int, input().split()))
for idx in idxs:
    idx -= 1
    grid[pos[idx][0]][pos[idx][1]] = 1
print(dfs(grid, 0, 0, 2 * R + 1))